lib/Parse/ParseDecl.cpp

```diff
@@ -5179,109 +5136,80 @@ ParserStatus Parser::parseTypeAttribute(TypeOrCustomAttr &result,
   llvm_unreachable("bad attribute kind");
 }
 
-static ParsedLifetimeDependenceKind getSILLifetimeDependenceKind(const Token &T) {
-  if (T.isContextualKeyword("_inherit")) {
-    return ParsedLifetimeDependenceKind::Inherit;
-  }
-  assert(T.isContextualKeyword("_scope"));
-  return ParsedLifetimeDependenceKind::Scope;
-}
-
-ParserStatus
-Parser::parseLifetimeEntries(SmallVectorImpl<LifetimeEntry> &specifierList) {
+ParserResult<LifetimeEntry> Parser::parseLifetimeEntry(SourceLoc loc) {
   ParserStatus status;
-  // TODO: Add fixits for diagnostics in this function.
-  do {
-    if (!isLifetimeDependenceToken()) {
-      break;
-    }
-
-    auto lifetimeDependenceKind = ParsedLifetimeDependenceKind::Default;
 
-    if (!isInSILMode()) {
-      // consume dependsOn
-      consumeToken();
-    } else {
-      lifetimeDependenceKind = getSILLifetimeDependenceKind(Tok);
-      // consume _inherit or _scope
-      consumeToken();
+  auto getLifetimeDependenceKind =
+      [&](Token Tok) -> std::optional<ParsedLifetimeDependenceKind> {
+    if (Tok.isContextualKeyword("copy") &&
+        peekToken().isAny(tok::identifier, tok::integer_literal,
+                          tok::kw_self)) {
+      return ParsedLifetimeDependenceKind::Inherit;
+    }
+    if (Tok.isContextualKeyword("borrow") &&
+        peekToken().isAny(tok::identifier, tok::integer_literal,
+                          tok::kw_self)) {
+      return ParsedLifetimeDependenceKind::Scope;
     }
+    return std::nullopt;
+  };
 
-    if (!Tok.isFollowingLParen()) {
-      diagnose(Tok, diag::expected_lparen_after_lifetime_dependence);
+  if (!Tok.isFollowingLParen()) {
+    diagnose(loc, diag::expected_lparen_after_lifetime_dependence);
+    status.setIsParseError();
+    return status;
+  }
+  // consume the l_paren
+  auto lParenLoc = consumeToken();
+
+  std::optional<LifetimeDescriptor> targetDescriptor;
+  if (Tok.isAny(tok::identifier, tok::integer_literal, tok::kw_self) &&
```

Comment:
I just want to clarify. Is `@lifetime(self: something)` and `@lifetime(1: something)` considered valid syntax? Because the latter won’t parse in the new parser since `1` is not a valid argument label.

Reply:
We were planning to use integer indices in the case of `dependsOn` especially for function types which may not have parameter names. `@lifetime(1: 0)` doesn't read as well as a positional `dependsOn(0)` would.  We likely can delete support for integer indices. I'll follow with a separate PR for this.

Reply:
I would be in favor of deleting integer target indices.